source/common/upstream/subset_lb.cc

```diff
@@ -477,20 +480,26 @@ void SubsetLoadBalancer::HostSubsetImpl::update(const HostVector& hosts_added,
   HostVectorSharedPtr hosts(new HostVector());
   HostVectorSharedPtr healthy_hosts(new HostVector());
 
+  // It's possible that hosts_added == original_host_set_.hosts(), e.g.: when
+  // calling refreshSubsets() if only metadata change. If so, we can avoid the
+  // predicate() call.
```

Comment:
Alternatively, I think you could change refreshSubsets to not pass the host set as hosts added since the underlying LBs don't need use information anyway.

Reply:
I would like to do that — it would improve things a lot — but we'd fail to enter this loop in `processSubsets`:

https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/subset_lb.cc#L200

... which means we wouldn't create the new subsets, if any. 

So we'd have to rework `processSubsets()` or also call `findOrCreateSubset()` from somewhere else... 

I'll look into it, but we can probably merge this in parallel since it's still better.